// @aeternity/aepp-calldata/cjs/main.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AciContractCallEncoder", {
  enumerable: true,
  get: function () {
    return _AciContractCallEncoder.default;
  }
});
Object.defineProperty(exports, "BytecodeContractCallEncoder", {
  enumerable: true,
  get: function () {
    return _BytecodeContractCallEncoder.default;
  }
});
Object.defineProperty(exports, "ContractByteArrayEncoder", {
  enumerable: true,
  get: function () {
    return _ContractByteArrayEncoder.default;
  }
});
Object.defineProperty(exports, "ContractEncoder", {
  enumerable: true,
  get: function () {
    return _ContractEncoder.default;
  }
});
Object.defineProperty(exports, "Encoder", {
  enumerable: true,
  get: function () {
    return _Encoder.default;
  }
});
Object.defineProperty(exports, "FateApiEncoder", {
  enumerable: true,
  get: function () {
    return _FateApiEncoder.default;
  }
});
Object.defineProperty(exports, "TypeResolver", {
  enumerable: true,
  get: function () {
    return _TypeResolver.default;
  }
});
exports.default = void 0;
var _Encoder = _interopRequireDefault(require("./api/Encoder.cjs"));
var _AciContractCallEncoder = _interopRequireDefault(require("./api/AciContractCallEncoder.cjs"));
var _BytecodeContractCallEncoder = _interopRequireDefault(require("./api/BytecodeContractCallEncoder.cjs"));
var _ContractByteArrayEncoder = _interopRequireDefault(require("./api/ContractByteArrayEncoder.cjs"));
var _FateApiEncoder = _interopRequireDefault(require("./api/FateApiEncoder.cjs"));
var _ContractEncoder = _interopRequireDefault(require("./api/ContractEncoder.cjs"));
var _TypeResolver = _interopRequireDefault(require("./api/TypeResolver.cjs"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * @deprecated use named exports instead
 */
var _default = exports.default = {
  Encoder: _Encoder.default,
  AciContractCallEncoder: _AciContractCallEncoder.default,
  BytecodeContractCallEncoder: _BytecodeContractCallEncoder.default,
  ContractByteArrayEncoder: _ContractByteArrayEncoder.default,
  FateApiEncoder: _FateApiEncoder.default,
  ContractEncoder: _ContractEncoder.default,
  TypeResolver: _TypeResolver.default
};
